### ISSUE TITLE
Inconsistent resolution of types for nested datatypes

### ISSUE THREAD

**Issue by moosbruggerj** [microsoft/pyright#4570]
**Describe the bug**
When defining a function to flatten an arbitrary deep nested list, the resulting type is different, depending whether the inner most type contains a dict with an union value. A plain union as innermost type does get correctly flattened. It only appears to be happening with a dict.

**To Reproduce**
see code below

**Expected behavior**
I would expect the type of `flat` to be `Sequence[Data]` and `Sequence[SimpleData]` respectively. Or at least exhibit the same flatten property.

**Screenshots or Code**
```python
from typing import TypeAlias, Sequence, TypeVar

T = TypeVar('T')

_NestedList: TypeAlias = T | Sequence['_NestedList[T]']
NestedList: TypeAlias = Sequence[_NestedList[T]]

Data: TypeAlias = dict[str, str | float]
SimpleData: TypeAlias = dict[str, str]

def flatten(value: NestedList[T]) -> Sequence[T]: ...

def test() -> Sequence[SimpleData]:
    value: Sequence[Sequence[SimpleData]] = [[{}], [{}]]
    flat =  flatten(value)
    reveal_type(flat)  # Sequence[dict[str, str]]
    return flat

def fail() -> Sequence[Data]:
    value: Sequence[Sequence[Data]] = [[{}], [{}]]
    flat =  flatten(value)
    reveal_type(flat)  # Sequence[list[dict[str, str | float]]]
    # expected: Sequence[dict[str, str | float]]
    return flat

def test2() -> Sequence[str | float]:
    value: Sequence[Sequence[str | float]] = [[""], [0.0]]
    flat =  flatten(value)
    reveal_type(flat)  # Sequence[str | float]
    return flat

```

**VS Code extension or command-line**
```
npx pyright@1.1.292 pyright_flatten.py
```

**Additional context**

**Comment by erictraut** [comment]
Thanks for the bug report. This turns out to be the same issue as one that was discussed recently on the [python/typing](https://github.com/python/typing/discussions/1342) discussion forum. Please see my response there.

This will be addressed in the next release.

**Comment by moosbruggerj** [comment]
Thank you for the quick fix and detailed explanation.

**Comment by erictraut** [comment]
This is included in pyright 1.1.293, which I just published. It will also be included in a future release of pylance.
